website_version_ce/models/experiment.py

```diff
@@ -0,0 +1,245 @@
+    a version in an experiment.
+    The googe_index is the index of a version in an experiment, used to send
+    data to Google Analytics.
+    """
+
+    _name = "website_version_ce.experiment.version"
+    _rec_name = "version_id"
+
+    version_id = fields.Many2one('website_version_ce.version',
+                                 string="Version", required=True,
+                                 ondelete='cascade')
+    experiment_id = fields.Many2one(
+        'website_version_ce.experiment',
+        string="Experiment",
+        required=True,
+        ondelete='cascade'
+    )
+    frequency = fields.Selection(
+        [('10', 'Less'), ('50', 'Medium'), ('80', 'More')],
+        string='Frequency', default='50')
+    google_index = fields.Integer(string='Google index')
+
+
+class Goals(models.Model):
+    """ Allow to define the goal of an experiment.
+    The goals are defined in the Google Analytics account and can be
+    synchronised in backend.
+    """
+    _name = "website_version_ce.goals"
+
+    name = fields.Char(string="Name", required=True)
+    google_ref = fields.Char(string="Google Reference", required=True)
+
+
+class Experiment(models.Model):
+    """ An experiment pointed to some experiment_versions and dispatch
+    each website visitor to a version.
+    """
+
+    _name = "website_version_ce.experiment"
+    _inherit = ['mail.thread']
+    _order = 'sequence'
+
+    @api.multi
+    @api.constrains('state')
+    def _check_view(self):
+        # No overlap for running experiments
+        for exp in self:
+            if exp.state == 'running':
+                for exp_ver in exp.experiment_version_ids:
+                    if exp_ver.search([
+                        ('version_id.view_ids.key', 'in',
+                         [v.key for v in exp_ver.version_id.view_ids]),
+                        ('experiment_id', '!=', exp_ver.experiment_id.id),
+                        ('experiment_id.website_id', '=',
+                         exp_ver.experiment_id.website_id.id),
+                        ('experiment_id.state', '=', 'running')
+                    ]):
+                        raise ValidationError(_(
+                            'This experiment contains a view which ' +
+                            'is already used in another running experience'))
+        return True
+
+    @api.multi
+    @api.constrains('website_id', 'experiment_version_ids')
+    def _check_website(self):
+        for exp in self:
+            for exp_ver in exp.experiment_version_ids:
+                if not exp_ver.version_id.website_id.id == exp.website_id.id:
+                    raise ValidationError(_(
+                        'This experiment must have versions which ' +
+                        'are in the same website'))
+
+    @api.model
+    def read_group(self, domain, fields, groupby, offset=0, limit=None,
+                   context=None, orderby=False, lazy=True):
```

Comment:
https://github.com/odoo/odoo/blob/9.0/openerp/models.py#L2017

I cannot change the way odoo defines its internal model functions.